### ISSUE TITLE
Inconsistent model behavior when loading model using load_state_dict in different ways

### ISSUE THREAD

**Issue by Rearcher** [pytorch/pytorch#23777]
## 🐛 Bug

I have a model which has two embedding layers. And I initialize the model, train it on GPU, and save its state dict to disk using `torch.save(model.state_dict, 'tmp.pt')`. 

And now I want to use previous model to do inference jobs, I find two ways below can cause different results:
1. initialize model on cpu -> load state dict -> model to gpu
2. initialize model on cpu -> model to gpu -> load state dict

**However, using just one embedding layer in my model will solve this.**

## To Reproduce

Run the code below, and it does these things:
1. fix random seeds
2. generate fake embedding weights and samples
3. initialize one model, train it, save it on disk
4. three different ways of doing inference, and the first  two asserts will fail

```python
import torch
import torch.nn as nn
import torch.optim as optim

def seed_torch(seed):
    import random
    random.seed(seed)
    
    import os
    os.environ['PYTHONHASHSEED'] = str(seed)
    
    import numpy as np
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed) # if you are using multi-GPU.
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.deterministic = True
    
seed_torch(2019)

class Model(nn.Module):
    def __init__(self, vectors):
        super(Model, self).__init__()
        self.emb = nn.Embedding.from_pretrained(
            vectors, freeze=False,
            padding_idx=0
        )

        # dummy embedding layers, just initialized and not used
        self.emb2 = nn.Embedding.from_pretrained(
            vectors, freeze=False,
            padding_idx=0
        )
        self.linear = nn.Linear(128, 2)
        
    def forward(self, s):
        s = self.emb(s)
        s = self.linear(s)
        s = s.sum(dim=1)
        return s
    
vectors = torch.randn(1000, 128)
# fake samples
train_s = torch.randint(1, 1000, size=(100, 30)).to('cuda')
train_y = torch.ones(100).to(dtype=torch.long).to('cuda')

# train
model = Model(vectors).to('cuda')
criterion = nn.CrossEntropyLoss(reduction='sum')
optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)

model.train()
model.zero_grad()
out = model(train_s)
loss = criterion(out, train_y)
loss.backward()
optimizer.step()

torch.save(model.state_dict(), 'tmp.pt')

# inference, model init -> to cuda -> load state dict
m1 = Model(vectors).to('cuda')
m1.load_state_dict(torch.load('tmp.pt'))
m1.eval()
with torch.no_grad():
    o1 = m1(train_s.to('cuda'))
    
# inference, model init -> load state dit -> to cuda
m2 = Model(vectors)
m2.load_state_dict(torch.load('tmp.pt'))
m2 = m2.to('cuda')
m2.eval()
with torch.no_grad():
    o2 = m2(train_s.to('cuda'))
    
# inference on cpu
m3 = Model(vectors)
m3.load_state_dict(torch.load('tmp.pt'))
m3.eval()
with torch.no_grad():
    o3 = m3(train_s.to('cpu'))
    
assert torch.allclose(o1, o2)
assert torch.allclose(o1.cpu(), o3)
assert torch.allclose(o2.cpu(), o3)
```

## Expected behavior
The assert in above code should all pass. 

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla K80
Nvidia driver version: 410.104
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip] numpy==1.16.2
[pip] pytorch-pretrained-bert==0.6.2
[pip] torch==1.1.0
[pip] torchvision==0.3.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-service               2.0.2            py36h7b6447c_0
[conda] mkl_fft                   1.0.12           py36ha843d7b_0
[conda] mkl_random                1.0.2            py36hd81dba3_0
[conda] pytorch                   1.1.0           py3.6_cuda10.0.130_cudnn7.5.1_0    pytorch
[conda] torchvision               0.3.0           py36_cu10.0.130_1    pytorch

**Comment by ezyang** [comment]
This seems like it should work. Marking as high priority to understand if there is just an error in the code or there is something serious going on.

**Comment by rgommers** [comment]
The differences are of order 1e-2. This doesn't look like a bug though. The root cause is that both embeddings use the same `vectors` input with `freeze=False`:
```
nn.Embedding.from_pretrained(
            vectors, freeze=False,
            padding_idx=0
```
The default for `freeze` is `True`; using that all asserts pass. Changing the use of the second embedding to use a unique tensor input:
```
        import copy
        self.emb2 = nn.Embedding.from_pretrained(
            copy.deepcopy(vectors), freeze=False,
            padding_idx=0
        )
```
also makes the asserts pass.

tl;dr the input vector gets modified in place, and since that's done in two separate embeddings, the results become inconsistent.

The documentation for `freeze` in `Embedding.from_pretrained` is clear about this, so I don't think there's anything to be done here.

**Comment by rgommers** [comment]
Okay I'll close this, since it seems pretty clear-cut. @ezyang @Rearcher please comment if you disagree.

**Comment by ezyang** [comment]
Thanks @rgommers!

**Comment by lgessler** [comment]
It's really puzzling to me that in the repro example, the asserts all pass if `self.emb2 = ...` is simply removed from the `Model`. This problem would be understandable to me if `self.emb2` were modified by autograd, but `self.emb2` is initialized and never used again. Can anyone give an intuition for what's going on here?
